### ISSUE TITLE
Unnecessary normalisation?

### ISSUE THREAD

**Issue by nad** [agda/agda#2907]
Consider the following code:
```agda
open import Agda.Builtin.Unit

id : (F : Set → Set) → ((A : Set) → F A) → ((A : Set) → F A)
id _ f = f

data Box (A : Set) : Set where
  [_] : A → Box A

{-# TERMINATING #-}

X : Set
X = Box X

ok : ⊤
ok = id _ (λ _ → tt) X

also-ok : ⊤
also-ok = id (λ _ → ⊤) (λ _ → _) X

loops : ⊤
loops = id _ (λ _ → _) X
```
The last expression triggers a stack overflow on my machine. I defined `X = Box X` to make sure that reduction of `X` to WHNF should not be problematic, so it seems to me as if Agda tries to normalise `X`. However, there is no need to normalise `X` here. Is Agda too strict?

I seem to recall reporting something similar in the past, but I failed to find any such issue.

**Comment by UlfNorell** [comment]
The culprit is `expandProjectedVars`:
https://github.com/agda/agda/blob/c2bf416f57f73594129141a105d35bd1c79ab7d1/src/full/Agda/TypeChecking/MetaVars.hs#L1008

@andreasabel: is normalisation necessary here?

**Comment by andreasabel** [comment]
Yes, we need to normalized meta arguments in the hope they become a variable, such that we can abstract. 
> The culprit is expandProjectedVars

I think that normalization was done even before this enhancement, see this comment:
https://github.com/agda/agda/blob/c2bf416f57f73594129141a105d35bd1c79ab7d1/src/full/Agda/TypeChecking/MetaVars.hs#L623-L627

I am not claiming this could not be done smarter.  For instance, one could have methods that try to normalize a term, but only spend N computation steps.  Then, one could simultaneously normalize all meta arguments, increasing N over time, until one finds a normal form that is definitely not a variable and will definitely place the constraint out of the pattern fragment.  Or, we could cap the time spent on normalization of the meta args and just give up if the time is used up.

However, we would need stronger reasons to act upon this than a lie about `{-# TERMINATING #-}`.

**Comment by nad** [comment]
> However, we would need stronger reasons to act upon this than a lie about `{-# TERMINATING #-}`.

The code that led me to report this issue contains an expensive but terminating piece of code, but I prefer non-terminating code in this kind of test case for two reasons:
* The code tends to be shorter.
* The code tends to be easier to test, there is no need for, say, a carefully tuned timeout value.

**Comment by andreasabel** [comment]
We could still have a special `try-to-normalize-and-eta-contract-to-a-variable-or-projection-thereof` function that would abort normalization early if there is no hope to arrive at the eta-expansion of a variable or projection.

**Comment by UlfNorell** [comment]
Is this not fixed by de4225eb?

**Comment by UlfNorell** [comment]
Closing, reopen as necessary.

**Comment by nad** [comment]
The test case in the OP still loops.

**Comment by UlfNorell** [comment]
It's now looping when pruning. More precisely here:

https://github.com/agda/agda/blob/32bc7b2fbe83ef88654028a21985da81b3bae86d/src/full/Agda/TypeChecking/MetaVars/Occurs.hs#L608-L612

This code was introduced to fix #1386.

I guess 1386 could be solved by being more conservative with pruning instead of more precise. I'm not sure how much more yellow that would result in. @andreasabel?

**Comment by andreasabel** [comment]
Indeed, my patch did not fix the problem, since pruning also normalizes unnecessarily.  I made an attempt at it during AIM XXVI but realized this requires more thorough thinking...

**Comment by jespercockx** [comment]
Perhaps the `isEtaVar` function could be used to avoid normalization here: https://github.com/agda/agda/blob/e34faff4485bd7055dd77d25c5707fe25b8b4856/src/full/Agda/TypeChecking/Records.hs#L814-L817

**Comment by jespercockx** [comment]
@andreasabel could you please either make a fix or bump this?
